File: documentation-examples-kotlin/src/kotlin/dev/typr/foundations/docs/routines/InoutProcedure.kt
package dev.typr.foundations.docs.routines

import dev.typr.foundationskt.*
import java.math.BigDecimal

@Suppress("unused")
class InoutProcedure {
    private val tx: Transactor? = null // placeholder

    //start
    companion object {
        // INOUT — the value goes in and comes back modified
        val applyDiscount =
            DbProcedure.define("apply_discount")
                .input(PgTypes.text)           // discount_code IN
                .inout(PgTypes.numeric)       // price INOUT — goes in, comes back modified
                .build()
    }

    fun applyDiscount(code: String, price: BigDecimal): BigDecimal =
        InoutProcedure.applyDiscount
            .call(code, price)
            .transact(tx!!)
    //stop
}

File: documentation-examples-scala/src/scala/dev/typr/foundationssc/docs/routines/InoutProcedure.scala
package dev.typr.foundationssc.docs.routines
import dev.typr.foundationssc.*

@SuppressWarnings(Array("unused"))
object InoutProcedure:
  private val tx: Transactor = null // placeholder

  // start
  // INOUT — the value goes in and comes back modified
  val applyDiscount: DbProcedure.Def2_1[String, BigDecimal, BigDecimal] =
    DbProcedure
      .define("apply_discount")
      .input(PgTypes.text) // discount_code IN
      .inout(PgTypes.numeric) // price INOUT
      .build()

  def applyDiscount(code: String, price: BigDecimal): BigDecimal =
    applyDiscount.call(code, price).transact(tx)
  // stop
